File: scripts/command-code.mjs
#!/usr/bin/env node

import { syncProviderModels } from "./model-registry.mjs";
import { fetchJson, fetchText, parseFlags, resolveModelsDir, logSyncResult, uniqueModelKey } from "./lib/shared.mjs";
import { stripParamInfoKey } from "./lib/clean-key.mjs";

const PROVIDER_NAME = "command_code";
const MODELS_API_URL = "https://api.commandcode.ai/provider/v1/models";
const PRICING_DOCS_URL = "https://commandcode.ai/docs/resources/pricing-limits";

// Command Code's "Go" tier ($1/mo plan) is restricted to open-source models.
// The live /provider/v1/models endpoint lists every model but carries no
// tier field, so the open-source set is resolved from the RSC payload in
// the pricing docs and intersected with the API by name.
const GO_TIER = "go";

// Minimum number of Go-tier models expected; guards against silent parsing breakage.
const MIN_EXPECTED_MODELS = 15;

function toModelKey(modelId) {
  const slashIndex = modelId.indexOf("/");
  const baseModelId = slashIndex === -1 ? modelId : modelId.slice(slashIndex + 1);

  return stripParamInfoKey(
    baseModelId
      .toLowerCase()
      .replace(/[:/]/g, "-")
      .replace(/[^a-zA-Z0-9._-]/g, "-")
      .replace(/-{2,}/g, "-")
  );
}

function extractOpenSourceModelNames(html) {
  const prefix = 'self.__next_f.push([1,"';
  let searchFrom = 0;

  while (true) {
    const startIdx = html.indexOf(prefix, searchFrom);
    if (startIdx === -1) break;

    const contentStart = startIdx + prefix.length;
    let i = contentStart;
    let payload = "";
    while (i < html.length) {
      if (html[i] === "\\" && i + 1 < html.length) {
        payload += html[i] + html[i + 1];
        i += 2;
      } else if (html.startsWith('"])', i)) {
        break;
      } else {
        payload += html[i];
        i++;
      }
    }

    if (payload.includes("opensource") && payload.includes("models")) {
      const unescaped = payload.replace(/\\"/g, '"');
      const bracketIdx = unescaped.indexOf("[");
      if (bracketIdx === -1) continue;

      const jsonStr = unescaped.slice(bracketIdx);
      // Trim trailing non-JSON content (e.g. trailing \n from RSC payload)
      const cleanJson = jsonStr.replace(/[^}\]]*$/, "");
      try {
        const data = JSON.parse(cleanJson);
        const modelObj = data[3];
        if (!modelObj || !modelObj.models) continue;

        const names = new Set();
        for (const m of modelObj.models) {
          if (m.category === "opensource" && m.name) {
            names.add(m.name);
          }
        }
        return names;
      } catch {
        continue;
      }
    }

    searchFrom = contentStart + payload.length;
  }

  throw new Error("Unable to locate open-source model data in Command Code pricing docs");
}

async function fetchGoTierModels() {
  const [apiPayload, docsHtml] = await Promise.all([
    fetchJson(MODELS_API_URL, { label: "Command Code /provider/v1/models" }),
    fetchText(PRICING_DOCS_URL, { label: "Command Code pricing docs", headers: { Accept: "text/html" } }),
  ]);

  if (!apiPayload || !Array.isArray(apiPayload.data)) {
    throw new Error("Unexpected Command Code /provider/v1/models payload format");
  }

  const openSourceNames = extractOpenSourceModelNames(docsHtml);
  if (openSourceNames.size === 0) {
    throw new Error("No open-source models found in Command Code pricing docs");
  }

  const goModels = apiPayload.data
    .filter((item) => item && typeof item === "object" && typeof item.id === "string" && typeof item.name === "string")
    .filter((item) => openSourceNames.has(item.name.trim()))
    .map((item) => ({ id: item.id.trim(), name: item.name.trim() }));

  if (goModels.length < MIN_EXPECTED_MODELS) {
    throw new Error(`Command Code resolved only ${goModels.length} Go-tier models (expected >= ${MIN_EXPECTED_MODELS})`);
  }

  return goModels;
}

function buildModelMap(models) {
  const map = new Map();
  const providerConfigByModel = new Map();

  for (const { id } of models) {
    const modelKey = uniqueModelKey(map, toModelKey(id), id);
    map.set(modelKey, id);
    providerConfigByModel.set(modelKey, { allowedTiers: [GO_TIER] });
  }

  return {
    modelMap: new Map([...map.entries()].sort(([a], [b]) => a.localeCompare(b))),
    providerConfigByModel,
  };
}

async function main() {
  const { dryRun } = parseFlags();
  const modelsDir = resolveModelsDir(import.meta.url);

  const models = await fetchGoTierModels();
  const { modelMap, providerConfigByModel } = buildModelMap(models);

  const result = syncProviderModels(modelsDir, PROVIDER_NAME, modelMap, {
    providerConfigByModel,
    managedProviderConfigKeys: ["allowedTiers"],
    dryRun,
  });

  logSyncResult({ label: "[command-code] Go-tier", count: modelMap.size, result, would: dryRun });
}

main().catch((error) => {
  console.error(error instanceof Error ? error.message : String(error));
  process.exitCode = 1;
});

File: packages/ai/pkg/sync/providers/commandcode.go
package providers

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"

	gosync "github.com/opendum/opendum/packages/ai/pkg/sync"
	"github.com/opendum/opendum/packages/ai/pkg/sync/httpfetch"
)

const (
	commandCodeModelsAPI  = "https://api.commandcode.ai/provider/v1/models"
	commandCodePricingURL = "https://commandcode.ai/docs/resources/pricing-limits"
	commandCodeMinModels  = 15
)

func commandCodeToModelKey(modelID string) string {
	base := modelID
	if idx := strings.Index(modelID, "/"); idx != -1 {
		base = modelID[idx+1:]
	}
	return stripKey(strings.ToLower(sanitizeKey(base)))
}

func commandCodeOpenSourceNames(html string) (map[string]bool, error) {
	const prefix = `self.__next_f.push([1,"`
	searchFrom := 0
	for {
		start := strings.Index(html[searchFrom:], prefix)
		if start == -1 {
			break
		}
		start += searchFrom
		contentStart := start + len(prefix)
		var payload strings.Builder
		i := contentStart
		for i < len(html) {
			if html[i] == '\\' && i+1 < len(html) {
				payload.WriteByte(html[i])
				payload.WriteByte(html[i+1])
				i += 2
			} else if strings.HasPrefix(html[i:], `"])`) {
				break
			} else {
				payload.WriteByte(html[i])
				i++
			}
		}
		raw := payload.String()
		if strings.Contains(raw, "opensource") && strings.Contains(raw, "models") {
			unescaped := strings.ReplaceAll(raw, `\"`, `"`)
			bracket := strings.Index(unescaped, "[")
			if bracket == -1 {
				searchFrom = contentStart + payload.Len()
				continue
			}
			jsonStr := unescaped[bracket:]
			jsonStr = strings.TrimRight(jsonStr, "\n")
			jsonStr = regexpTrimTrailing(jsonStr)
			var data []any
			if err := json.Unmarshal([]byte(jsonStr), &data); err != nil {
				searchFrom = contentStart + payload.Len()
				continue
			}
			if len(data) > 3 {
				if obj, ok := data[3].(map[string]any); ok {
					if models, ok := obj["models"].([]any); ok {
						names := map[string]bool{}
						for _, m := range models {
							mm, ok := m.(map[string]any)
							if !ok {
								continue
							}
							if mm["category"] == "opensource" {
								if name, ok := mm["name"].(string); ok && name != "" {
									names[name] = true
								}
							}
						}
						return names, nil
					}
				}
			}
		}
		searchFrom = contentStart + payload.Len()
	}
	return nil, fmt.Errorf("Unable to locate open-source model data in Command Code pricing docs")
}

func regexpTrimTrailing(s string) string {
	i := len(s)
	for i > 0 && s[i-1] != '}' && s[i-1] != ']' {
		i--
	}
	return s[:i]
}

func SyncCommandCode(ctx context.Context, modelsDir string) (gosync.ProviderResult, error) {
	client := &http.Client{Timeout: 20 * time.Second}
	var apiPayload struct {
		Data []struct {
			ID   string `json:"id"`
			Name string `json:"name"`
		} `json:"data"`
	}
	var docsHTML string
	var wg sync.WaitGroup
	var apiErr, docsErr error
	wg.Add(2)
	go func() {
		defer wg.Done()
		apiErr = httpfetch.FetchJSON(ctx, client, commandCodeModelsAPI, &apiPayload, &httpfetch.Options{Label: "Command Code /provider/v1/models"})
	}()
	go func() {
		defer wg.Done()
		docsHTML, docsErr = httpfetch.FetchText(ctx, client, commandCodePricingURL, &httpfetch.Options{Label: "Command Code pricing docs", Headers: map[string]string{"Accept": "text/html"}})
	}()
	wg.Wait()
	if apiErr != nil {
		return gosync.ProviderResult{Provider: "command_code"}, apiErr
	}
	if docsErr != nil {
		return gosync.ProviderResult{Provider: "command_code"}, docsErr
	}
	openSource, err := commandCodeOpenSourceNames(docsHTML)
	if err != nil {
		return gosync.ProviderResult{Provider: "command_code"}, err
	}
	if len(openSource) == 0 {
		return gosync.ProviderResult{Provider: "command_code"}, fmt.Errorf("No open-source models found in Command Code pricing docs")
	}
	type model struct{ id string }
	models := []model{}
	for _, item := range apiPayload.Data {
		id := strings.TrimSpace(item.ID)
		name := strings.TrimSpace(item.Name)
		if id == "" || name == "" {
			continue
		}
		if !openSource[name] {
			continue
		}
		models = append(models, model{id})
	}
	if len(models) < commandCodeMinModels {
		return gosync.ProviderResult{Provider: "command_code"}, fmt.Errorf("Command Code resolved only %d Go-tier models (expected >= %d)", len(models), commandCodeMinModels)
	}
	modelMap := map[string]string{}
	extra := map[string]map[string]any{}
	for _, m := range models {
		base := commandCodeToModelKey(m.id)
		key := base
		suffix := 2
		for {
			existing, ok := modelMap[key]
			if !ok || existing == m.id {
				break
			}
			key = fmt.Sprintf("%s-%d", base, suffix)
			suffix++
		}
		modelMap[key] = m.id
		extra[key] = map[string]any{"allowedTiers": []any{"go"}}
	}
	modelMap = sortMap(modelMap)
	res, err := gosync.SyncProviderRegistry(modelsDir, "command_code", modelMap, extra, []string{"allowedTiers"})
	if err != nil {
		return res, err
	}
	if len(res.Added) == 0 && len(res.Removed) == 0 && len(res.Updated) == 0 {
		fmt.Printf("Command Code Go-tier models are already up to date (%d models).\n", len(modelMap))
	} else {
		fmt.Printf("Command Code: %d models (added %d, removed %d, updated %d).\n", len(modelMap), len(res.Added), len(res.Removed), len(res.Updated))
	}
	return res, nil
}
